### PR TITLE
switch plotting code to makie

### PR DESCRIPTION
**Pull request by SimonDanisch**
- [x] port all recipes & tests
- [x] documentation
- [x] percy visual regression tests 
- [ ] ~~improved themeabiltity~~ will do this in another PR

# Visual changes from this PR (first image -> this PR, second image -> master)

## All together 1
all_together_makie
all_together_plots

## All together 2
all_together_2_makie
all_together_2_plots

## Confusion matrix :Column
confusion_col_makie
confusion_col_plots

## Confusion matrix :Row 
confusion_row_makie
confusion_row_plots

## Kappas IRA
kappas_ira_makie
kappas_ira_plots

## Kappas no IRA
kappas_no_ira_makie
kappas_no_ira_plots

## PR Curves
pr_makie
pr_plots

## PRG curves
prg_makie
prg_plots

## Reliability calibration
reliability_calibration_makie
reliability_calibration_plots

## ROC Curves
roc_makie
roc_plots

### CODE REVIEW

**Comment by SimonDanisch** [comment]
Alright, a preview of the docs for the plotting functions is up at:
https://beacon-biosignals.github.io/Lighthouse.jl/previews/PR29/plotting/

@hannahilea, maybe you can push to this PR to fill in some more details about what those plotting functions do (if desired^^) ? Those details should go into the docstrings here:
https://github.com/beacon-biosignals/Lighthouse.jl/blob/470d7ef67432bf10227a8da3d4956f1c1d154dfc/src/plotting.jl#L348

File: src/plotting.jl
Comment (ericphanson) [inline src/plotting.jl:53]: typo?

**Comment by hannahilea** [comment]
> @hannahilea, maybe you can push to this PR to fill in some more details about what those plotting functions do (if desired^^)

Let's not block this PR on docs, since we already know those are not in great shape in this repo and adding them isn't top priority (yet).

**Comment by SimonDanisch** [comment]
I tried a few things to inline the plots somewhere, but it wasn't that easy without uploading the images anywhere...
So I guess for the comparison of Plots.jl + Makie.jl plots, one has to download this HTML file:
https://gist.github.com/SimonDanisch/1eec17096c9bef5637b103761c29edd1#file-plots_vs_makie-html

File: src/plotting.jl
Comment (SimonDanisch) [reply src/plotting.jl]: Woops, some German sneaked in^^

**Comment by ericphanson** [comment]
that comparison looks good! My thoughts:

* The text in the Makie ones seems a lot more legible, with the white text when the color is dark, and also on the bars in the barplots in a much more readable position.
* I think the legend in `all_together_2` in the Makie would might work better as having 2 or 3 columns instead of 1, but that's kind of a minor thing (and it's definitely better than the weird empty axis in the Plots one!).
* most of the lines in the plots between the Plots vs Makie seem the same, but there are some differences in the numbers in the legends, like the AUC values in many of the plots, and in `kappas_no_ira` the "Multiclass" one is much larger for Makie (0.047 vs 0.011). Is that expected or is there a bug?

**Comment by SimonDanisch** [comment]
Hm, I let this run multiple times and got pretty different values every time, which is indeed surprising considering that the plots do look pretty much like a 1:1 match.
I also experimented with using StableRNGs to no avail. 
This is the output after running `]test` 2 times:
all_together_2
all_together_2

This will of course also make percy tests hard... @hannahilea do you think we are missing restting an RNG somewhere, or is the training and metrics output hard to make fully deterministic?

**Comment by hannahilea** [comment]
> @hannahilea do you think we are missing resetting an RNG somewhere

Yep! We use the global rng in `majority(...)`, which means we should reseed the global rng at the start of testing:
```
Random.seed!(0)
```

**Comment by SimonDanisch** [comment]
I checked that one explicitly and it looks like we use a local RNG in the tests:
https://github.com/beacon-biosignals/Lighthouse.jl/blob/sd/makie/test/learn.jl#L93

**Comment by hannahilea** [comment]
`majority(...)` is called by the default `elected` param in `learn(...)`, so tests that call `learn(...)` without passing in a non-default `elected` arg end up using that default rng. Looks like that is happening at least here [1], hence needing the global rng reseeded.


[1] https://github.com/beacon-biosignals/Lighthouse.jl/blob/66cbc09977cb04f0dd6d277ce1e5db5832664b4d/test/learn.jl#L47

**Comment by SimonDanisch** [comment]
Oh ha, of course you're right :) I thought I looked at all call sites of majority and made sure RNGs get passed through...

File: Project.toml
Comment (hannahilea) [inline Project.toml:18]: Need a compat bound for `AbstractPlotting`

File: docs/make.jl
Comment (hannahilea) [inline docs/make.jl:10]: Yay :)

**Comment by hannahilea** [comment]
It's an easy one to miss...I was thrown by some nondeterminism in the LighthouseFlux tests that was due to this callsite, so I'm well familiar with it 😅

File: docs/src/plotting.md
Comment (hannahilea) [inline docs/src/plotting.md:15]: Since (as a user) it's easy to accidentally invert the labels on the x and y axes, it might be nice if the example use-case started from labeled data, to demonstrate expected behavior; if so, could construct a confusion matrix via Lighthouse calls/labeled data.

File: docs/src/plotting.md
Comment (hannahilea) [inline docs/src/plotting.md:59]: this should maybe be a StableRNG (or the global rng should be seeded at the start of the docs) to make sure docs always generate the same plots?

File: docs/src/plotting.md
Comment (hannahilea) [inline docs/src/plotting.md:89]: ditto rng

Comment (hannahilea) [reply docs/src/plotting.md:89]: (also several other places!)

File: docs/src/plotting.md
Comment (hannahilea) [inline docs/src/plotting.md:146]: Oh geez, haha! We should remove these (in a future PR is fine) 
```suggestion
# These are currently not used in plotting, but are still passed to `plot_binary_discrimination_calibration_curves`!
```

File: test/runtests.jl
Comment (hannahilea) [inline test/runtests.jl:14]:
```suggestion
mkpath(plot_results)
```

File: Project.toml
Comment (hannahilea) [inline Project.toml:4]:
```suggestion
version = "0.12.0"
```

File: src/plotting.jl
Comment (hannahilea) [inline src/plotting.jl:92]: It feels a little weird for this to call `plot_pr_curves!` under the hood; maybe indicates that `plot_pr_curves!(...)` should be renamed to something more generic? e.g. a `plot_curves!(...)` function that both `plot_roc_curves!` and `plot_pr_curves!` call.

Comment (hannahilea) [reply src/plotting.jl:92]: ...now seeing that it also gets called by `plot_reliability_calibration_curves!(...)` below, and possibly others, so I'm even more in favor of a generic name for the shared functionality.

Comment (SimonDanisch) [reply src/plotting.jl:92]: Yes, thanks for the reminder...That was a stopgap solution to try out if I can use the same function for the others, and then never came up with a good name for it. Lets rename it and make `plot_pr_curves` call it!

File: src/plotting.jl
Comment (hannahilea) [inline src/plotting.jl:163]: :|

Comment (SimonDanisch) [reply src/plotting.jl:163]: On the upside, this is fixed in the newest release ;)

File: src/plotting.jl
Comment (hannahilea) [inline src/plotting.jl:188]: oh interesting, I haven't seen this used reversed before.

File: src/plotting.jl
Comment (hannahilea) [inline src/plotting.jl:177]: Is this comment still valid?

**Comment by hannahilea** [comment]
Cool! So far so good---I'm extra excited for the good documentation and examples with images, THANK YOU for adding that!!

I'm going to hold off on reviewing the rendered plots themselves until we have side-by-side comparisons of old vs new for the same data (i.e. once you've done the reseed before generating both sets of plots), since (of course) the most important sanity check is that the plotted values haven't changed at all for the same underlying data.

**Review by hannahilea — CHANGES_REQUESTED** [review]
...as noted above (I _always_ forget to save my comments for the "finish your review" box), will review the visual changes once we have data parity between the before/after plots.

File: test/runtests.jl
Comment (SimonDanisch) [reply test/runtests.jl]: I actually just added `isdir(plot_results) ? rm(plot_results; force=true, recursive=true) : mkdir(plot_results)`, to make sure we don't have any old pictures in there ...

Comment (SimonDanisch) [reply test/runtests.jl]: I just added this:
```julia
# Remove any old plots
isdir(plot_results) && rm(plot_results; force=true, recursive=true)
mkdir(plot_results)
```
To make sure we don't work with old plots...

Comment (hannahilea) [reply test/runtests.jl]: Nice!

File: docs/src/plotting.md
Comment (SimonDanisch) [reply docs/src/plotting.md:15]: Good point! 
Can you detail how you would improve the example?

**Comment by SimonDanisch** [comment]
Alright, the rng in majority did the trick :) 
Here is the updated comparison:
https://gist.github.com/SimonDanisch/1eec17096c9bef5637b103761c29edd1/archive/2668d2e4c8f528b59d96da4d5fcd6b02719774df.zip

File: docs/src/plotting.md
Comment (hannahilea) [reply docs/src/plotting.md:18]: something like
```
classes = ["red", "orange", "yellow", "green"]
ground_truth = [1, 1, 1, 2, 2, 2, 3, 3, 3, 4, 4, 4]
predicted_labels = [1, 1, 1, 1, 2, 2, 4, 4, 4, 4, 4, 1]
confusion = confusion_matrix(length(classes), zip(predicted_labels, ground_truth))
fig, ax, p = plot_confusion_matrix(confusion, classes, :Row)
```

**Comment by hannahilea** [comment]
Ahhhh Simon this is looking so good! :D 

Looking at the side-by-side plots, here are a few things I think need tweaking before we're good to go:
- [ ]  **kappas_ira.png** I think the plot classes are inverted---the labels vs the colors are switched, so previously the grey bars were the expert-expert and now they are the algorithm-vs-expert. 
- [ ] **kappas_ira.png** Could we make sure that the expert-expert bars are above the expert-algorithm bars (as in the original version)? This is intentional, to match the data formatting of publications that we want to easily compare against.
- [ ] **kappas_ira.png** The title should be the same as the old title
- [ ]  **kappas_ira.png** Looks like we've lost the grid/minor x ticks; probably want those back pending intentional design choices around internal styling. 
- [ ]  **reliability_calibration.png** Would it be easy to make the scatter plot point outlines be the color of the line rather than black? The black is a little distracting.
- [ ] **all_together.png/all_together_2.png** The titles of some of the component plots are centered rather than left-justified; they were originally left-justified to make it easier to differentiate from x-axis labels.
- [ ] **Missing plot: binary_discrimination_calibration_curves** Looks like this one just doesn't exist as a solo comparison plot or in the `all_together.png` plots

Also, style improvements that can absolutely wait until future PRs, unless they're easy and you want to: 
- all individual plots (ROC, PR, calibration, confusion) would ideally be square
 
Also, if at all possible, would be nice to have the side-by-side comparison images in this PR directly, rather than a link-out, so that we don't lose the history if you delete your gist at some point, and so that users can see the changes in this PR to know what they're getting when they bump versions. Makes sense to wait until it's the final version, though, if you prefer!

**Comment by hannahilea** [comment]
> rng doesn't broadcast on 1.3

So I _think_ if you wanted you could instead bump the required Julia version to 1.6. We're already going to be bumping the Lighthouse minor version, and I don't think any internal Beacon users are using <1.6 at this point.... Your call!

**Comment by SimonDanisch** [comment]
* The missing decorations in kappas were intentional to remove noise, considering that we already have the labels. I added them back for now, but didn't update the x ticks yet (so its still just 0 + 1).

I will try to figure out how to best upload a report here, so its easier to view... Its somehow surprisingly hard to just upload an html file somewhere and be able to directly look at it.

File: test/learn.jl
Comment (hannahilea) [inline test/learn.jl:258]:
```suggestion
        plot_data = last(logger.logged["validation_set_evaluation/metrics_per_epoch"])
```

...to get the discrimination validation plot to show up in the grouped plots! (Somehow between when this plot and its group test were added and now, this change got lost---it isn't a regression due to this PR, but making this change will add slight coverage going forward!

Comment (hannahilea) [reply test/learn.jl]: ...this was totally a regression caused by me in #22, whoops! If only we'd had percy then....

**Comment by SimonDanisch** [comment]
Alright, new comparison: https://serene-jang-51d2d2.netlify.app/
Note, sharpness comparisons aren't very useful with this, since those are PNGs and they're not shown at their native resolution...
Before we merge, I can collect the comparison into the PR description, so that one can see the differences right away ;)

File: src/plotting.jl
Comment (hannahilea) [inline src/plotting.jl:140]:
```suggestion
                         scatter=(markershape=markershape, markersize=markersize, strokewidth=0))
```
...I think this is the change to remove the circle outlines on the discrimination calibration plot, so that it matches the reliability calibration plot?

**Review by hannahilea — APPROVED** [review]
Barring the one tiny suggested change, this looks good to go!! 

> Before we merge, I can collect the comparison into the PR description, so that one can see the differences right away ;)

Sounds good to me!

File: src/plotting.jl
Comment (SimonDanisch) [reply src/plotting.jl:140]: Good catch, thank you! :)

**Review by ericphanson — APPROVED** [review]
looks great!

**Pull request merged**
